core/circuit_breaker/circuit_breaker.go

```diff
@@ -0,0 +1,214 @@
+package circuit_breaker
+
+import (
+	"github.com/alibaba/sentinel-golang/core/base"
+	"github.com/alibaba/sentinel-golang/core/stat"
+	"github.com/alibaba/sentinel-golang/util"
+	"sync/atomic"
+	"time"
+)
+
+type CircuitBreaker interface {
+	getRule() BreakerRule
+
+	CanPass(ctx *base.EntryContext, node base.StatNode, acquireCount uint32) *base.TokenResult
+}
+
+// average rt circuit breaker will cut resource if the rt of resource exceed the threshold of rule.
+type averageRtCircuitBreaker struct {
+	// status of the breaker
+	cut    util.AtomicBool
+	rule   *averageRtBreakerRule
+	metric base.ReadStat
+}
+
+func newAverageRtCircuitBreaker(rule *averageRtBreakerRule) *averageRtCircuitBreaker {
+	resNode := stat.GetResourceNode(rule.Resource)
+	metric := resNode.GetOrCreateSlidingWindowMetric(rule.SampleCount, rule.IntervalInMs)
+	return &averageRtCircuitBreaker{
+		rule:   rule,
+		metric: metric,
+	}
+}
+
+func newAverageRtCircuitBreakerWithMetric(rule *averageRtBreakerRule, metric base.ReadStat) *averageRtCircuitBreaker {
+	return &averageRtCircuitBreaker{
+		rule:   rule,
+		metric: metric,
+	}
+}
+
+func (b averageRtCircuitBreaker) getRule() BreakerRule {
+	return b.rule
+}
+
+func (b *averageRtCircuitBreaker) CanPass(_ *base.EntryContext, _ base.StatNode, _ uint32) *base.TokenResult {
+	// currently, the breaker is in silent status, direct return blocked .
+	if b.cut.Get() {
+		return base.NewTokenResultBlocked(base.BlockTypeCircuitBreaking, "CircuitBreaking")
+	}
+	rule := b.rule
+	if rule == nil {
+		return base.NewTokenResultPass()
+	}
+
+	if b.metric == nil {
+		logger.Errorf("Metric of averageRtCircuitBreaker is nil")
+		return base.NewTokenResultPass()
+	}
+
+	avgRt := b.metric.AvgRT()
+	if avgRt < rule.Threshold {
+		atomic.StoreInt64(&rule.PassCount, 0)
+		return base.NewTokenResultPass()
+	}
+	if util.IncrementAndGetInt64(&rule.PassCount) < rule.RtSlowRequestAmount {
+		return base.NewTokenResultPass()
+	}
+	// trigger circuit breaker
+	if b.cut.CompareAndSet(false, true) {
+		go util.RunWithRecover(func() {
+			// recover after TimeSilent seconds
+			time.Sleep(time.Second * time.Duration(rule.TimeSilent))
+			atomic.StoreInt64(&rule.PassCount, 0)
+			b.cut.Set(false)
+		}, logger)
+	}
+	return base.NewTokenResultBlocked(base.BlockTypeCircuitBreaking, "CircuitBreaking")
+}
+
+// error ratio circuit breaker will cut resource if the error ratio of resource exceed the threshold of rule.
+type errorRatioCircuitBreaker struct {
+	// status of the breaker
+	cut    util.AtomicBool
+	rule   *errorRatioBreakerRule
+	metric base.ReadStat
+}
+
+func newErrorRatioCircuitBreaker(rule *errorRatioBreakerRule) *errorRatioCircuitBreaker {
+	resNode := stat.GetResourceNode(rule.Resource)
+	metric := resNode.GetOrCreateSlidingWindowMetric(rule.SampleCount, rule.IntervalInMs)
```

Comment:
What if the `resNode` is nil? (i.e. when the resource has not been visited yet)

Note: This problem could be eliminated in later versions, as we're planning to create individual stat structures for circuit breakers, rather than use the universal ResourceNode.